File: 2024/Soohyeun/2024-12-01-CoinChange.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # using dp
        # create an array, length of amount + 1
        # go throught the array and create nested iteration with coins, and check [index-coin] and find minimum

        memo = [float('inf')] * (amount + 1)
        memo[0] = 0

        for i in range(1, amount+1):
            for coin in coins:
                prev = i - coin
                if prev >= 0 and memo[prev] != float('inf'):
                    memo[i] = min(memo[prev]+1, memo[i])

        return memo[amount] if memo[amount] != float('inf') else -1

File: 2024/YounSangHo/Week10/2024-12-01-coin-change.js
class Solution {
  /**
   * @param {number[]} coins
   * @param {number} amount
   * @return {number}
   */
  coinChange(coins, amount) {
    let memo = {};

    const dfs = (amount) => {
      if (amount === 0) return 0;
      if (memo[amount] !== undefined) return memo[amount];

      let res = Infinity;
      for (let coin of coins) {
        if (amount - coin >= 0) {
          res = Math.min(res, 1 + dfs(amount - coin));
        }
      }

      memo[amount] = res;
      return res;
    };

    const minCoins = dfs(amount);
    return minCoins === Infinity ? -1 : minCoins;
  }
}
